messagequeue/sqs/client.go

```diff
@@ -0,0 +1,209 @@
+package sqs
+
+import (
+	"net/url"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/beaconsoftwarellc/gadget/v2/errors"
+	"github.com/beaconsoftwarellc/gadget/v2/log"
+	"github.com/beaconsoftwarellc/gadget/v2/messagequeue"
+)
+
+const (
+	serviceAttributeName   = "service"
+	methodAttributeName    = "method"
+	awsTraceHeaderName     = "AWSTraceHeader"
+	maxWaitTime            = 20 * time.Second
+	maxMessageDequeueCount = 10
+)
+
+// VisibilityTimeout should be used to timeout the context for messages
+// 		- Functions that handle messages MUST respect the timeout
+// You can extend the timeout by calling 'ChangeMessageVisibility' and setting it
+//		higher than what it was. This can be done by the worker currently holding
+//		the message
+// Best Practice is Queue per Producer
+
+// SQS interface for sending and receiving messages from a simple queueing service
+// instance.
+type Client interface {
+	// Enqueue the passed message
+	Enqueue(m *messagequeue.Message) error
+	// Enqueue all the passed messages as a batch
+	EnqueueBatch(messages []*messagequeue.Message) error
+	// Dequeue up to the passed count of messages waiting up to the passed
+	// duration
+	Dequeue(count int, wait time.Duration) ([]*messagequeue.Message, error)
+	// Delete the passed message from the queue so that it is not processed by
+	// other workers
+	Delete(*messagequeue.Message) error
+}
+
+// New SQS instance located at the passed URL
+func New(queueLocator *url.URL) Client {
+	return &sdk{
+		queueUrl: queueLocator,
+	}
+}
+
+type sdk struct {
+	queueUrl *url.URL
+	api      API
+}
+
+func (mq *sdk) API() (API, error) {
+	if nil != mq.api {
+		return mq.api, nil
+	}
+	var (
+		err  error
+		sess *session.Session
+	)
+	sess, err = session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if nil == err {
+		mq.api = sqs.New(sess)
+	}
+	return mq.api, err
+}
+
+func (mq *sdk) Enqueue(msg *messagequeue.Message) error {
```

Comment:
It would be easier to have a single
`func (mq *sdk) Enqueue(msg *messagequeue.Message...) error` which wraps `api.SendMessageBatch`.

Reply:
I updated Batch to return per item so this is OBE.